### ISSUE TITLE
QuickScreenshot plugin

### ISSUE THREAD

**Issue by bhavinmoriya** [xournalpp/xournalpp#2787]
**Affects versions :**
 - OS: Ubuntu 20.4
 - (Linux only) Desktop environment: Wayland, 
 - Which version of libgtk do you use
     - can be obtained by running `xournalpp --version` or in the **About** window
 - Version of Xournal++: Xournal++ 1.1.0+dev 
└──libgtk: 3.24.20
 - Installation method: using deb file

**Describe the bug**
I am trying to use QuickScreenshot plugin. I do have scrot installed, and it works totally fine. Now, when I enable QuickScreenshot plugin the selection prompt appears which prompt the saving window after selection of area. I did give a name to the file and saved it in a folder. But when I access the folder there is no file there. I did start Xournal++ from terminal before doing all this and the message prompted on terminal could be see here, https://www.dropbox.com/s/owkt6azanytxp1u/Xournalpp.png?dl=0

**To Reproduce**
1. Step 1 : Open Xournal++ from terminal
2. Step 2 Enable QuickScreenshot plugin
3. Select the portion to be screenshoted
4. Name of the file and location 
5. Check the saved file (In my case there was no file)

**Expected behavior**
I expect it to screenshot the region selected.

**Screenshots of Problem** 
(Optional) If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other information or context about the problem here.

**Comment by rolandlo** [comment]
@bhavinmoriya Please backup the `main.lua` file in the plugin folder (`/usr/share/xournalpp/plugins/QuickScreenshot/` under normal circumstances) and replace it by the m̀ain.lua` file in this archive:  
[main.zip](https://github.com/xournalpp/xournalpp/files/5952661/main.zip)

Then run the plugin and report the console messages. When you see the temporary filename displayed in the console  (before the Save As dialog is shown) you may also check if it exists.

**Comment by bhavinmoriya** [comment]
> @bhavinmoriya Please backup the `main.lua` file in the plugin folder (`/usr/share/xournalpp/plugins/QuickScreenshot/` under normal circumstances) and replace it by the m̀ain.lua` file in this archive:
> [main.zip](https://github.com/xournalpp/xournalpp/files/5952661/main.zip)
> 
> Then run the plugin and report the console messages. When you see the temporary filename displayed in the console (before the Save As dialog is shown) you may also check if it exists.

Hi Roland,
I did do exactly as you said but no success yet. While I launched Xournal++ the message appeared is 1. Enabling plugin I got 2 and then saving the screenshot I got the 3. I did try to save the file with another name but still no success as in 4.
Thanks a lot for helping me out here.

PS. Reading the message in terminal as in 3 I went to /tmp/ and there was an image there with the strange name as shown in the image I sent to you. So I do have image just that not in a place I wanted it to be saved and without the name I wanted to give.

**Comment by rolandlo** [comment]
When you see that the temporary file exists, it could still be empty.

As your console messages show you get a `giblib error: no image grabbed`. According to this [Stackoverflow answer](https://stackoverflow.com/questions/26326664/scrot-giblib-error-saving-to-file-failed) the problem could be that the temporary file generated has no file name extension. So please try replacing the line 
```lua
       local tmpFilename = os.tmpname()
 ```
in `main.lua` by

```lua
       local tmpFilename = os.tmpname() .. ".png"
 ```
and see if it works now.

**Comment by bhavinmoriya** [comment]
I did change as recommended but still only available file is in tmp folder. Terminal does emit the success message as before but while I go to the respected place there is no file.

![Screenshot from 2021-02-16 09-11-41](https://user-images.githubusercontent.com/15716951/108061429-10a9ae80-7037-11eb-8c71-77ee3ddec9f5.png)

**Comment by rolandlo** [comment]
Can you try to open the temporary file and see if the image is alright? There is no error message anymore, so it's a bit strange, that it still doesn't work.

**Comment by rolandlo** [comment]
To check whether the renaming is the problem: Please replace the lines
```lua
        os.rename(tmpFilename, filename)
        foundUtility = true
        print("Success!")
 ```
in `main.lua` by

```lua
        success, error = os.rename(tmpFilename, filename)
        if (success) then
          foundUtility = true
          print("Success!")
        else
          print("Renaming failed with error message: " .. error)
        end
 ```

**Comment by bhavinmoriya** [comment]
> Can you try to open the temporary file and see if the image is alright? There is no error message anymore, so it's a bit strange, that it still doesn't work.

Image is absolutely alright and there is no message while I open it using eog.

**Comment by bhavinmoriya** [comment]
> To check whether the renaming is the problem: Please replace the lines
> 
> ```lua
>         os.rename(tmpFilename, filename)
>         foundUtility = true
>         print("Success!")
> ```
> 
> in `main.lua` by
> 
> ```lua
>         success, error = os.rename(tmpFilename, filename)
>         if (success) then
>           foundUtility = true
>           print("Success!")
>         else
>           print("Renaming failed with error message: " .. error)
>         end
> ```

I did the requested changes and got the following message in the terminal.

![Screenshot from 2021-02-16 15-06-04](https://user-images.githubusercontent.com/15716951/108103523-d48f4180-7068-11eb-993c-650bc2f87473.png)

**Comment by rolandlo** [comment]
Ok, seems we found the problem. "Invalid cross-device link" means that your temporary folder and the home folder are on different partitions (and different file systems?). That's why the renaming fails. Could you please tell about the partitions and file systems you use for /home and /tmp?
I am not sure how to solve it yet. For your purposes it would be enough to replace
 ```lua
         os.rename(tmpFilename, filename)
 ```
 
 in `main.lua` by
 
 ```lua
         os.execute("mv " .. tmpFilename .. " " .. filename)
 ```
This is a solution which won't work on Windows, for example.

**Comment by bhavinmoriya** [comment]
> Ok, seems we found the problem. "Invalid cross-device link" means that your temporary folder and the home folder are on different partitions (and different file systems?). That's why the renaming fails. Could you please tell about the partitions and file systems you use for /home and /tmp?
> I am not sure how to solve it yet. For your purposes it would be enough to replace
> 
> ```lua
>         os.rename(tmpFilename, filename)
> ```
> 
> in `main.lua` by
> 
> ```lua
>         os.execute("mv " .. tmpFilename .. " " .. filename)
> ```
> 
> This is a solution which won't work on Windows, for example.

os.rename(tmpFilename, filename) appears in comment and the modified chunk of the code. Should I uncomment the respective part and do the suggested change or otherwise? Following is the change that I have done as per your previous suggestion,

success, error = os.rename(tmpFilename, filename)
        if (success) then
          foundUtility = true
          print("Success!")
        else
          print("Renaming failed with error message: " .. error)
        end
        ---os.rename(tmpFilename, filename)
        ---foundUtility = true
        ---print("Success!")

Regarding the fs, /tmp and /home has the same filesystem as I see here in this image,

![filesy](https://user-images.githubusercontent.com/15716951/108125056-af113080-7086-11eb-8a03-16ea224d4454.png)

**Comment by rolandlo** [comment]
Ok, your /tmp folder and /home folder are on two different partitions with the same file system (ext4). 

Sorry for being unclear about the change. You should remove (or turn into comments) the modified chunk (7 lines) and uncomment the other 3 lines and apply the new change there.

**Comment by bhavinmoriya** [comment]
I guess I was not clever enough to understand what you have been saying.
Now, it is working like charm.
You made me cross the sea :)
Thanks a lot Roland!

Bhavin Moriya <https://sites.google.com/site/bhavinmoriya/>

Assistant Professor

The Federal University of Viçosa - Brazil.

Departamento de Matemática <http://www.dma.ufv.br/>

Av. P.H. Rolfs, s/n

36570-000 – Viçosa.

Minas Gerais

Brasil.

    <http://www.linkedin.com/in/bhavin-moriya-b0b88b2>
<https://livelifemaverick.blogspot.com/>
<https://www.youtube.com/channel/UCkK-IRTdBcnWcrwBImzlRyw>

"Only the gentle are ever really strong."

- James Dean

**Comment by bhavinmoriya** [comment]
> Ok, your /tmp folder and /home folder are on two different partitions with the same file system (ext4).
> 
> Sorry for being unclear about the change. You should remove (or turn into comments) the modified chunk (7 lines) and uncomment the other 3 lines and apply the new change there.

I guess I was not clever enough to understand what you have been saying. Now, it is working like charm. 
You made me cross the sea :)
Thanks a lot Roland!

**Comment by bhavinmoriya** [comment]
Shall I close the issue?

**Comment by rolandlo** [comment]
No, please leave it open, as we will have to fix it in a PR and will have to do it slightly differently if aiming to eventually have the plugin on all operating systems. Thanks for assisting with debugging.

**Comment by bhavinmoriya** [comment]
Just for curiosity, what exactly is PR?

Bhavin Moriya <https://sites.google.com/site/bhavinmoriya/>

Assistant Professor

The Federal University of Viçosa - Brazil.

Departamento de Matemática <http://www.dma.ufv.br/>

Av. P.H. Rolfs, s/n

36570-000 – Viçosa.

Minas Gerais

Brasil.

    <http://www.linkedin.com/in/bhavin-moriya-b0b88b2>
<https://livelifemaverick.blogspot.com/>
<https://www.youtube.com/channel/UCkK-IRTdBcnWcrwBImzlRyw>

"Only the gentle are ever really strong."

- James Dean

**Comment by bhavinmoriya** [comment]
It was a great pleasure assisting you.

Bhavin Moriya <https://sites.google.com/site/bhavinmoriya/>

Assistant Professor

The Federal University of Viçosa - Brazil.

Departamento de Matemática <http://www.dma.ufv.br/>

Av. P.H. Rolfs, s/n

36570-000 – Viçosa.

Minas Gerais

Brasil.

    <http://www.linkedin.com/in/bhavin-moriya-b0b88b2>
<https://livelifemaverick.blogspot.com/>
<https://www.youtube.com/channel/UCkK-IRTdBcnWcrwBImzlRyw>

"Only the gentle are ever really strong."

- James Dean

**Comment by rolandlo** [comment]
PR = pull request
As we often talk about "pull requests" we mostly abbreviate it like that.

**Comment by bhavinmoriya** [comment]
Thanks.

Bhavin Moriya <https://sites.google.com/site/bhavinmoriya/>

Assistant Professor

The Federal University of Viçosa - Brazil.

Departamento de Matemática <http://www.dma.ufv.br/>

Av. P.H. Rolfs, s/n

36570-000 – Viçosa.

Minas Gerais

Brasil.

    <http://www.linkedin.com/in/bhavin-moriya-b0b88b2>
<https://livelifemaverick.blogspot.com/>
<https://www.youtube.com/channel/UCkK-IRTdBcnWcrwBImzlRyw>

"Only the gentle are ever really strong."

- James Dean

**Comment by createyourpersonalaccount** [comment]
Thank you for this thorough discussion. There are two bugs discovered:
1) The file is always saved as PNG regardless of what the user wants. 
2) There's some issue with `os.rename()` across partitions. 

For 1), we can change that with a configuration option for the plugin in the top down menu. For 2) I don't quite understand it, I think the Lua implementation doesn't handle moving files correctly and it needs to be substituted by a custom solution that will again differ across operating systems.

**Comment by createyourpersonalaccount** [comment]
I think I have a solution in mind that will cover most cases; it will be a custom copy function that uses `std::filesystem`. If it's fine with everyone I will work on this and submit a PR soon.

**Comment by rolandlo** [comment]
@createyourpersonalaccount Sure, you are very welcome to submit a PR.

**Comment by createyourpersonalaccount** [comment]
I have uploaded the changes in [changes.zip](https://github.com/xournalpp/xournalpp/files/6035957/changes.zip). I can't figure out how to do a PR with these changes, and github is always a stumbling block for me. It will probably take me 2-3 hours to figure this out... If someone wants to do the work of submitting this PR, please be my guest, I am sorry but it is impossible for me to understand github...

I am including the changes:

`plugins/QuickScreenshot/main.lua:`
1. Changed the comment description of `go()` to match its behavior.
2. Changed `os.rename()` to `app.glib_rename()` that takes care of the different partition issue.
`src/plugin/luaapi_application.h:`
1. Included the `<gtk/gtk.h>` header to use glib's functions, in particular `g_file_move()`.
2. Implemented `applib_glib_rename()` for use by Lua.
3. Fixed `applib_saveAs`: It used to always suggest `Untitled.png` as a filename, but it is a generic function not tied to images.

**Comment by rolandlo** [comment]
@createyourpersonalaccount  Thanks for the fix. As suggested by you I created a pull request for your changes, see #2864. Could you please check if I didn't miss any changes? It seems to work fine, but I haven't tested it moving across partitions yet.

**Comment by rolandlo** [comment]
Ok, I have now tested it saving the screenshot on a USB stick. This worked fine. It doesn't without the fix.

**Comment by createyourpersonalaccount** [comment]
Thank you @rolandlo, I appreciate your help. I'll try to submit a PR myself next time, I am too scatter brained right now.

**Comment by rolandlo** [comment]
> As you notice, I fix the renaming issue, but I have not explicitly allowed the user to choose between PNG and other file names. Scrot takes care of things on its own, but other programs may not... Simply choosing `foo.JPG` as a filename right now works fine.

That's not quite true. As `scrot` saves to the temporary file, it saves it in `png` format (which is its default). It can get **renamed** to `.jpg`, but that doesn't make it a valid `.jpg`-file. It's just a `.png`-file with a `.jpg`-ending. If we want to fix it (and allow `.jpg` and `.gif` as alternatives) we either let the user decide about the file format **before** creating the screenshot (which requires one more click) or use a convert tool after the image has been created (which sounds like the better solution to me). Nevertheless if the demand is not too high, I would rather stick with only supporting `.png`. If some user wishes, he can always add a `os.execute` statement calling the conversion tool in the plugin.

**Comment by rolandlo** [comment]
This was fixed in #2864.
